bakerytest/bakerytest.go

```diff
@@ -127,3 +131,174 @@ func (d *Discharger) PublicKeyForLocation(loc string) (*bakery.PublicKey, error)
+		waiting: map[string]discharge{},
+	}
+	d.Mux.Handle("/visit", visitHandler)
+	d.Mux.Handle("/wait", http.HandlerFunc(d.wait))
+	server := httptest.NewTLSServer(d.Mux)
+	svc, err := bakery.NewService(bakery.NewServiceParams{
+		Location: server.URL,
+		Locator:  locator,
+	})
+	if err != nil {
+		panic(err)
+	}
+	httpbakery.AddDischargeHandler(d.Mux, "/", svc, d.checker)
+	startSkipVerify()
+	d.Discharger = Discharger{
+		Service: svc,
+		server:  server,
+	}
+	return d
+}
+
+func (d *InteractiveDischarger) checker(req *http.Request, cavId, cav string) ([]checkers.Caveat, error) {
+	d.mu.Lock()
+	id := fmt.Sprintf("%d", d.id)
+	d.id++
+	d.waiting[id] = discharge{cavId, make(chan error, 1)}
+	d.mu.Unlock()
+	visitURL := fmt.Sprintf("%s/visit?waitid=%s", d.Discharger.server.URL, id)
+	waitURL := fmt.Sprintf("%s/wait?waitid=%s", d.Discharger.server.URL, id)
+	return nil, httpbakery.NewInteractionRequiredError(visitURL, waitURL, nil, req)
+}
+
+func (d *InteractiveDischarger) wait(w http.ResponseWriter, r *http.Request) {
+	r.ParseForm()
+	d.mu.Lock()
+	discharge, ok := d.waiting[r.Form.Get("waitid")]
+	d.mu.Unlock()
+	if !ok {
+		code, body := httpbakery.ErrorToResponse(errgo.Newf("invalid waitid %q", r.Form.Get("waitid")))
+		httprequest.WriteJSON(w, code, body)
+		return
+	}
+	defer func() {
+		d.mu.Lock()
+		delete(d.waiting, r.Form.Get("waitid"))
+		d.mu.Unlock()
+	}()
+	var err error
+	select {
+	case e := <-discharge.c:
+		err = e
+	case <-time.After(5 * time.Minute):
+		code, body := httpbakery.ErrorToResponse(errgo.New("timeout waiting for interaction to complete"))
+		httprequest.WriteJSON(w, code, body)
+		return
+	}
+	if err != nil {
+		code, body := httpbakery.ErrorToResponse(err)
+		httprequest.WriteJSON(w, code, body)
+		return
+	}
+	m, err := d.Service.Discharge(
+		bakery.ThirdPartyCheckerFunc(
+			func(cavId, caveat string) ([]checkers.Caveat, error) {
+				return nil, nil
+			},
+		),
+		discharge.cavId,
+	)
+	if err != nil {
+		code, body := httpbakery.ErrorToResponse(err)
+		httprequest.WriteJSON(w, code, body)
+		return
+	}
+	httprequest.WriteJSON(
+		w,
+		http.StatusOK,
+		httpbakery.WaitResponse{
+			Macaroon: m,
+		},
+	)
+}
+
+// FinishInteraction signals to the InteractiveDischarger that a
+// particular interaction is complete. It causes any waiting requests to
+// return. If err is not nil then it will be returned by the
+// corresponding /wait request.
+func (d *InteractiveDischarger) FinishInteraction(w http.ResponseWriter, r *http.Request, err error) {
+	r.ParseForm()
+	d.mu.Lock()
+	discharge, ok := d.waiting[r.Form.Get("waitid")]
+	d.mu.Unlock()
+	if !ok {
+		code, body := httpbakery.ErrorToResponse(errgo.Newf("invalid waitid %q", r.Form.Get("waitid")))
+		httprequest.WriteJSON(w, code, body)
+		return
+	}
+	select {
+	case discharge.c <- err:
+	default:
+		panic("cannot finish interaction " + r.Form.Get("waitid"))
+	}
+	return
+}
+
+// URL generates a URL addressed path that will be handled correctly by
```

Comment:
"URL addressed path" reads oddly to me.

// URL returns a URL addressed to the given path in the discharger
// that contains any discharger context information found
// in the given request. Use this to generate intermediate URLs
// before calling FinishInteraction.

?